File: source code/practice/src/main/java/com/decade/practice/utils/ChatUtils.java
package com.decade.practice.utils;

import com.decade.practice.persistence.jpa.embeddables.ChatIdentifier;
import com.decade.practice.persistence.jpa.entities.Chat;
import com.decade.practice.persistence.jpa.entities.User;

import java.util.UUID;

public class ChatUtils {


    public static User inspectPartner(Chat chat, User me) {
        return chat.getFirstUser().equals(me) ? chat.getSecondUser() : chat.getFirstUser();
    }

    public static UUID inspectPartner(ChatIdentifier chat, UUID me) {
        return chat.getFirstUser().equals(me) ? chat.getSecondUser() : chat.getFirstUser();
    }

}

File: practice/src/main/java/com/decade/practice/util/ChatUtils.kt
package com.decade.practice.util

import com.decade.practice.model.embeddable.ChatIdentifier
import com.decade.practice.model.entity.Chat
import com.decade.practice.model.entity.User
import java.util.*

fun Chat.inspectOwner(me: UUID): User {
    return if (identifier.firstUser == me) firstUser else secondUser
}

fun Chat.inspectOwner(me: User): User {
    return if (firstUser == me) firstUser else secondUser
}

fun Chat.inspectOwner(username: String): User {
    return if (firstUser.username == username)
        firstUser
    else
        secondUser
}

// TODO: convert inspections to extension methods
fun Chat.inspectPartner(me: User): User {
    return if (firstUser == me) secondUser else firstUser
}

fun Chat.inspectPartner(me: UUID): User {
    return if (identifier.firstUser == me) secondUser else firstUser
}

fun toIdentifier(u1: UUID, u2: UUID): ChatIdentifier {
    return if (u1 < u2) ChatIdentifier(u1, u2) else ChatIdentifier(u2, u1)
}